File: frontend/src/lib/observation-colors.ts
/**
 * Single source of truth for pest / disease colours across the SPA.
 *
 * Resolution order (highest precedence first):
 *   1. Live override map fetched from the Pest / Plant Disease doctypes
 *      (their ``pests_legend_color`` / ``disease_legend_color`` fields).
 *      Edit a doc → cache invalidates server-side → next call re-fetches.
 *   2. Canonical hardcoded fallback (kept in sync with the backend
 *      ``observation_colors.PEST_DEFAULTS`` / ``DISEASE_DEFAULTS``) so a
 *      fresh install — or a name we haven't seen before but matches a
 *      synonym — still renders something meaningful.
 *   3. Neutral grey when nothing matches.
 *
 * Use the React hook ``useObservationColors`` from a component to get a
 * resolver bound to the live override map; non-React callers can call
 * ``pestColor(name)`` / ``diseaseColor(name)`` directly and they'll use
 * whatever the most recent fetch loaded into the module-level cache.
 */

import { useEffect, useState } from "react";
import { call } from "./frappe";

export type ObsKind = "pest" | "disease";

export interface ObsColorMap {
  pests: Record<string, string>;
  diseases: Record<string, string>;
}

const norm = (s: string): string =>
  String(s || "")
    .toLowerCase()
    .trim()
    .replace(/\s+/g, " ");

/** ──────────────────────────────────────────────────────────────────────
 *  Canonical hardcoded fallback (mirror of observation_colors.py).
 *  ────────────────────────────────────────────────────────────────────── */

export const PEST_PALETTE: Record<string, string> = {
  FCM: "#dc2626",
  Helicoverpa: "#eab308",
  Spodoptera: "#f97316",
  Duponchelia: "#38bdf8",
  Thrips: "#2563eb",
  Spidermites: "#8b4513",
  Aphids: "#16a34a",
  "White Flies": "#6b7280",
  Mealybugs: "#d4a017",
};

export const DISEASE_PALETTE: Record<string, string> = {
  "Powdery Mildew": "#166534",
  "Downy Mildew": "#ec4899",
  Botrytis: "#a855f7",
  Agrobacteria: "#eab308",
  Rust: "#c8a165",
};

export const OBS_DEFAULT_COLOR = "#9ca3af";

const PEST_LOOKUP: Record<string, string> = (() => {
  const m: Record<string, string> = {};
  const add = (keys: string[], hex: string) =>
    keys.forEach((k) => {
      m[norm(k)] = hex;
    });
  add(["FCM", "False Codling Moth", "False Codling Moths"], "#dc2626");
  add(["Helicoverpa", "Helicoverpa Armigera"], "#eab308");
  add(["Spodoptera", "Spodoptera Litura", "Spodoptera Frugiperda"], "#f97316");
  add(
    ["Duponchelia", "Duponchella", "Duponchelia Fovealis"],
    "#38bdf8",
  );
  add(["Thrips"], "#2563eb");
  add(
    ["Spidermites", "Spidermite", "Spider Mite", "Spider Mites", "Red Spider Mite"],
    "#8b4513",
  );
  add(["Aphids", "Aphid"], "#16a34a");
  add(["White Flies", "White Fly", "Whitefly", "Whiteflies"], "#6b7280");
  add(["Mealybugs", "Mealybug", "Mealy Bug", "Mealy Bugs"], "#d4a017");
  add(["Unidentified Moth", "Unknown Moth"], "#fb923c");
  return m;
})();

const DISEASE_LOOKUP: Record<string, string> = (() => {
  const m: Record<string, string> = {};
  const add = (keys: string[], hex: string) =>
    keys.forEach((k) => {
      m[norm(k)] = hex;
    });
  add(["Powdery Mildew"], "#166534");
  add(["Downy Mildew"], "#ec4899");
  add(["Botrytis", "Botyrtis"], "#a855f7");
  add(["Agrobacteria", "Agrobacterium"], "#eab308");
  add(["Rust"], "#c8a165");
  return m;
})();

/** ──────────────────────────────────────────────────────────────────────
 *  Live doctype-fed override map. Module-level cache so every call reuses
 *  one in-flight fetch.
 *  ────────────────────────────────────────────────────────────────────── */

let liveColors: ObsColorMap = { pests: {}, diseases: {} };
let liveColorsPromise: Promise<ObsColorMap> | null = null;
type Subscriber = (map: ObsColorMap) => void;
const subscribers = new Set<Subscriber>();

function emitLive() {
  subscribers.forEach((fn) => {
    try {
      fn(liveColors);
    } catch {
      /* never let a subscriber break the others */
    }
  });
}

/** Fire-and-forget fetch of the doctype-stored colours. Cached: a second
 *  caller during the in-flight request piggybacks on the same promise. */
export function loadObservationColors(force = false): Promise<ObsColorMap> {
  if (!force && liveColorsPromise) return liveColorsPromise;
  liveColorsPromise = call<ObsColorMap>(
    "upande_scp.serverscripts.observation_colors.get_observation_colors",
    {},
  )
    .then((r) => {
      liveColors = {
        pests: (r && r.pests) || {},
        diseases: (r && r.diseases) || {},
      };
      emitLive();
      return liveColors;
    })
    .catch(() => liveColors);
  return liveColorsPromise;
}

/** Subscribe to live-map updates. Returns an unsubscribe fn. */
export function subscribeObservationColors(fn: Subscriber): () => void {
  subscribers.add(fn);
  return () => {
    subscribers.delete(fn);
  };
}

/** Snapshot of the most recent live map (or empty if not yet loaded). */
export function getObservationColors(): ObsColorMap {
  return liveColors;
}

/** ──────────────────────────────────────────────────────────────────────
 *  Resolvers.
 *  ────────────────────────────────────────────────────────────────────── */

export function pestColor(
  name: string,
  override?: Record<string, string>,
): string {
  if (!name) return OBS_DEFAULT_COLOR;
  const live = override ?? liveColors.pests;
  if (live?.[name]) return live[name];
  // Some upstream callers normalise to lowercase before persisting; also try
  // a case-insensitive match against the live map.
  const k = norm(name);
  if (live) {
    for (const [doc, hex] of Object.entries(live)) {
      if (norm(doc) === k) return hex;
    }
  }
  const hit = PEST_LOOKUP[k];
  return hit || OBS_DEFAULT_COLOR;
}

export function diseaseColor(
  name: string,
  override?: Record<string, string>,
): string {
  if (!name) return OBS_DEFAULT_COLOR;
  const live = override ?? liveColors.diseases;
  if (live?.[name]) return live[name];
  const k = norm(name);
  if (live) {
    for (const [doc, hex] of Object.entries(live)) {
      if (norm(doc) === k) return hex;
    }
  }
  const hit = DISEASE_LOOKUP[k];
  return hit || OBS_DEFAULT_COLOR;
}

export function observationColor(
  name: string,
  kind: ObsKind,
  overrides?: Partial<ObsColorMap>,
): string {
  return kind === "pest"
    ? pestColor(name, overrides?.pests)
    : diseaseColor(name, overrides?.diseases);
}

/** Pick a readable text colour (white or charcoal) for a given hex fill. */
export function readableInk(hexBackground: string): string {
  const m = /^#?([a-f0-9]{6})$/i.exec(hexBackground.replace("#", ""));
  if (!m) return "#1f2937";
  const n = parseInt(m[1], 16);
  const r = (n >> 16) & 0xff;
  const g = (n >> 8) & 0xff;
  const b = n & 0xff;
  const L = (0.2126 * r + 0.7152 * g + 0.0722 * b) / 255;
  return L > 0.62 ? "#1f2937" : "#ffffff";
}

/** ──────────────────────────────────────────────────────────────────────
 *  React integration.
 *  ────────────────────────────────────────────────────────────────────── */

/** Returns the live doctype colour map plus a re-render guarantee — the
 *  hook subscribes to updates so a slow first fetch refreshes consumers. */
export function useObservationColors(): {
  map: ObsColorMap;
  pest: (name: string) => string;
  disease: (name: string) => string;
} {
  const [map, setMap] = useState<ObsColorMap>(liveColors);
  useEffect(() => {
    void loadObservationColors();
    return subscribeObservationColors(setMap);
  }, []);
  return {
    map,
    pest: (name: string) => pestColor(name, map.pests),
    disease: (name: string) => diseaseColor(name, map.diseases),
  };
}

/** Static legend entries — used by the colour-key panels. */
export const PEST_LEGEND: Array<{ name: string; color: string }> = Object.entries(
  PEST_PALETTE,
).map(([name, color]) => ({ name, color }));

export const DISEASE_LEGEND: Array<{ name: string; color: string }> =
  Object.entries(DISEASE_PALETTE).map(([name, color]) => ({ name, color }));

File: upande_scp/serverscripts/scouting/observation_colors.py
"""Observation colour resolver.

Storage lives on the Pest and Plant Disease doctypes — `pests_legend_color`
and `disease_legend_color`. This module:

  - exposes a tiny whitelisted endpoint the SPA calls once on boot, so the
    same map drives every legend, marker, and badge;
  - holds the canonical defaults the system was specced with so a fresh
    install / new pest gets a sensible colour without manual seeding.

The cache invalidation hooks live in cache_utils.py — editing a Pest or
Plant Disease doc busts ``K_PEST_COLORS`` / ``K_DISEASE_COLORS`` so the
next call rebuilds the map.
"""

import frappe

from upande_scp.serverscripts.scouting.get_complete_scouting_entries import (
    _cached_disease_colors,
    _cached_pest_colors,
)


# Canonical colour map. Used both for first-time seeding (see
# ``seed_canonical_colors``) and as a fallback in ``get_observation_colors``
# when a doc exists but its colour field is empty. Keep keys aligned with
# the doc ``name`` (autoname is ``common_name``), but match case-insensitively
# at lookup time so naming-rule drift doesn't silently lose a pest.
PEST_DEFAULTS = {
    "FCM": "#dc2626",            # Red
    "Helicoverpa": "#eab308",    # Yellow
    "Spodoptera": "#f97316",     # Orange
    "Duponchelia": "#38bdf8",    # Sky Blue
    "Duponchella": "#38bdf8",    # alt spelling encountered in production data
    "Thrips": "#2563eb",         # Blue
    "Spidermites": "#8b4513",    # Brown
    "Spider Mite": "#8b4513",
    "Spider Mites": "#8b4513",
    "Aphids": "#16a34a",         # Green
    "White Flies": "#6b7280",    # Grey
    "Whitefly": "#6b7280",
    "Mealybugs": "#d4a017",      # Gold
    "Mealybug": "#d4a017",
    # Moth observations seen in raw data but not formally listed —
    # share Spodoptera's moth-orange family so the marker remains readable.
    "Unidentified Moth": "#fb923c",
    # --- multi-crop expansion (Rose / Avocado / Coffee) ---
    "Scale Insects": "#78716b",
    "Weevils": "#57534e",
    "Caterpillars": "#84cc16",
    "Coconut Bug": "#b45309",
    "Fruit fly (Bactocera)": "#0891b2",
    "Fruit fly (Ceratitis)": "#0e7490",
    "Leaf Rollers": "#65a30d",
    "Loopers": "#4d7c0f",
    "Mosquito Bugs": "#be123c",
    "Stinkbug": "#15803d",
    "Unidentified Insects": "#9ca3af",
    # Coffee pests
    "Antestia Bug": "#7c3aed",
    "Capsid Bug": "#db2777",
    "Lace Bug": "#0d9488",
    "Leaf Skeletonizer": "#ca8a04",
    "Leaf Miner": "#a16207",
    "Tailed Caterpillar": "#22c55e",
    "Systates Weevil": "#78350f",
    "Kenya Mealybug": "#facc15",
    "Brown Scale": "#92400e",
    "Yellow Termites": "#fde047",
    "Green Scale": "#4ade80",
    "Coffee Thrips": "#1d4ed8",
    "Berry Moth": "#e11d48",
    "Berry Borer": "#831843",
    "Sting Caterpillar": "#a3e635",
}

DISEASE_DEFAULTS = {
    "Powdery Mildew": "#166534",  # Dark Green
    "Downy Mildew": "#ec4899",    # Pink
    "Botrytis": "#a855f7",        # Purple
    "Botyrtis": "#a855f7",        # alt spelling encountered in production data
    "Agrobacteria": "#eab308",    # Yellow
    "Agrobacterium": "#eab308",
    "Rust": "#c8a165",            # Tan
    # --- multi-crop expansion (Rose / Avocado / Coffee) ---
    "Bacterial Wilt": "#b91c1c",
    # Coffee diseases
    "Coffee Berry Disease": "#7f1d1d",
    "Coffee Leaf Rust": "#ea580c",
    "Coffee Wilt (Fusarium)": "#713f12",
    "Brown Eye Spot": "#a16207",
    "Bacterial Blight of Coffee": "#1e3a8a",
    # Avocado diseases
    "Anthracnose": "#292524",
    "Cercospora Spot": "#6d28d9",
    "Phytophthora Root Rot": "#365314",
    "Avocado Scab": "#9a3412",
}


def _normalize(s):
    return (s or "").strip().lower()


def _canonical_pest_color(name):
    key = _normalize(name)
    for k, v in PEST_DEFAULTS.items():
        if _normalize(k) == key:
            return v
    return None


def _canonical_disease_color(name):
    key = _normalize(name)
    for k, v in DISEASE_DEFAULTS.items():
        if _normalize(k) == key:
            return v
    return None


@frappe.whitelist(allow_guest=False)
def get_observation_colors():
    """Return ``{pests: {name: hex}, diseases: {name: hex}}``.

    Doctype-stored colour wins; if a doc has the field blank, the canonical
    default fills in so legends never render grey for a known pest.
    """
    pests = {}
    for row in _cached_pest_colors() or []:
        name = row.get("name")
        if not name:
            continue
        hex_value = row.get("pests_legend_color") or _canonical_pest_color(name)
        if hex_value:
            pests[name] = hex_value

    diseases = {}
    for row in _cached_disease_colors() or []:
        name = row.get("name")
        if not name:
            continue
        hex_value = row.get("disease_legend_color") or _canonical_disease_color(name)
        if hex_value:
            diseases[name] = hex_value

    return {"pests": pests, "diseases": diseases}


@frappe.whitelist()
def seed_canonical_colors(force=False):
    """Idempotently populate the canonical colour for every Pest /
    Plant Disease record listed in the defaults above.

    By default only fills empty colour fields so operator overrides are
    preserved; pass ``force=1`` to overwrite every value.
    """
    force_flag = str(force).lower() in ("1", "true", "yes")
    updated = {"pests": [], "diseases": []}

    for name, hex_value in PEST_DEFAULTS.items():
        if not frappe.db.exists("Pest", name):
            continue
        current = frappe.db.get_value("Pest", name, "pests_legend_color")
        if current and not force_flag:
            continue
        frappe.db.set_value("Pest", name, "pests_legend_color", hex_value)
        updated["pests"].append(name)

    for name, hex_value in DISEASE_DEFAULTS.items():
        if not frappe.db.exists("Plant Disease", name):
            continue
        current = frappe.db.get_value("Plant Disease", name, "disease_legend_color")
        if current and not force_flag:
            continue
        frappe.db.set_value("Plant Disease", name, "disease_legend_color", hex_value)
        updated["diseases"].append(name)

    if updated["pests"] or updated["diseases"]:
        frappe.cache().delete_value("scouting_dashboard:pest_colors")
        frappe.cache().delete_value("scouting_dashboard:disease_colors")
        frappe.db.commit()

    return updated


def after_migrate():
    """Hook target — run automatically on ``bench migrate`` so a fresh
    deployment never renders a known pest in fallback grey.
    """
    try:
        seed_canonical_colors()
    except Exception:
        # Don't fail the migration over a colour seed.
        frappe.logger().exception("upande_scp: observation_colors.after_migrate failed")
